### PR TITLE
Fix payment plan webhook handling and refactor to use existing invoice and payment tables

### PR DESCRIPTION
**Pull request by dwenderf**
# Refactor payment plans to xero_payments architecture
## Summary

This PR refactors the payment plan system from a separate `payment_plans` table to using the existing `xero_payments` table architecture. This consolidation simplifies the data model, improves maintainability, and fixes several critical bugs discovered during the refactoring process.

### 🏗️ Architecture Refactor

**Migration to xero_payments table:**

- Consolidated payment plan data into `xero_payments` table using `payment_type` ('installment' vs 'full')

- Added `installment_number` and `planned_payment_date` columns

- Created `payment_plan_summary` view for aggregated queries

- Restricted view access to service_role only (used by admin APIs and cron jobs)

 

**Payment status flow:**

- `staged` → Initial state when payment plan is created

- `pending` → First payment ready to process

- `planned` → Future payments waiting for their scheduled date

- `processing` → Currently being charged

- `synced` → Successfully synced to Xero

- `cancelled` → Superseded by early payoff

- `failed` → Payment failed after max attempts

 

### 🐛 Critical Bug Fixes

 

**1. Accounting Accuracy - Installment Rounding**

- **Problem:** Math.round() could cause sum of installments to not equal invoice total

- **Example:** $100.01 ÷ 4 = 4 × $25.00 = $100.00 (lost 1 cent)

- **Solution:** Last installment absorbs rounding remainder ($25.00 + $25.00 + $25.00 + $25.01 = $100.01)

- **Test Coverage:** 6 comprehensive tests validate perfect totals for all edge cases

 

**2. Early Payoff Reliability & Accounting**

- **Problem 1:** Synchronous charge → DB update could lose payment on server crash

- **Problem 2:** Would create 3 xero_payments ($2500 each) but only 1 Stripe payment ($7500)

- **Solution:** Webhook-based flow with 'cancelled' status

  1. Cancel planned payments (set `sync_status = 'cancelled'`)

  2. Create single staged payment for full remaining balance

  3. Create payment record & charge Stripe

  4. Webhook updates staged payment to pending (guaranteed even on crash)

 

**3. Schema Mismatches - contact_id Column**

- Fixed 3 locations using non-existent `contact_id` column:

  - `hasOutstandingBalance()` in payment-plan-service

  - `getTotalOutstandingBalance()` in payment-plan-service

  - Early payoff API endpoint

- Now correctly extract from `staging_metadata->>'user_id'`

 

**4. Payment Plan Eligibility Check**

- Added missing `payment_plan_enabled` flag check in `canUserCreatePaymentPlan()`

- Users now require both admin approval AND saved payment method

 

### 🔧 Code Quality Improvements

 

**DRY Principle - Centralized Configuration:**

- Created `payment-plan-config.ts` with shared constants:

  - `PAYMENT_PLAN_INSTALLMENTS = 4`

  - `INSTALLMENT_INTERVAL_DAYS = 30`

  - `MAX_PAYMENT_ATTEMPTS = 3`

  - `RETRY_INTERVAL_HOURS = 24`

- Replaced all hardcoded values across 5 files

- Removed duplicate `MAX_PAYMENT_ATTEMPTS` definition

 

**Code Consolidation:**

- Extracted duplicate webhook payment status update logic into `updatePaymentPlanStatuses()` helper

- Removed hardcoded 4-payment check (now flexible for any installment count)

- Applied object shorthand syntax

 

**Null Safety:**

- Added null safety to 3 `staging_metadata` spread operations

- Prevents crashes when metadata is missing

 

**Performance:**

- Fixed N+1 query in `getUserPaymentPlans()` using joins instead of loops

- Single query now fetches payment plans with registration data

 

**Permissions:**

- Fixed permission mismatch: admin endpoints now use `createAdminClient()` to query restricted views

- Updated RLS policies to match actual access patterns

 

### 📝 Documentation

 

- Clarified comments about payment status transitions

- Updated comments to match implementation (payment_plan_enabled check)

- Added inline documentation for complex logic

 

### ✅ Testing

 

**New Tests:**

- Created `payment-plan-service.test.ts` with 6 tests for installment distribution

- Tests cover perfect divisions, rounding errors, edge cases, and metadata

 

**Fixed Tests:**

- Updated `user-payment-plans.test.ts` to mock `createAdminClient()`

- Updated early payoff test expectations for new error messages

- All payment plan tests updated for xero_payments architecture

 

**Results:**

- ✅ All 56 tests passing across 8 test suites

 

## Database Changes

 

**New Migration:** `2025-11-06-refactor-payment-plans-to-xero-payments.sql`

- Adds new columns to `xero_payments` table

- Creates `payment_plan_summary` view

- Drops old `payment_plans` and `payment_plan_transactions` tables

- Data migration skipped (development environment)

 

### Core Service Layer

- `src/lib/services/payment-plan-service.ts` - Complete rewrite for xero_payments architecture

- `src/lib/services/payment-plan-config.ts` - **NEW** Centralized configuration

 

### API Routes

- `src/app/api/stripe-webhook/route.ts` - Updated for new architecture + early payoff webhook

- `src/app/api/cron/payment-plans/route.ts` - Process scheduled payments from xero_payments

- `src/app/api/user/payment-plans/early-payoff/route.ts` - Webhook-based early payoff flow

- `src/app/api/admin/payment-plans/route.ts` - Query payment_plan_summary view

- `src/app/api/admin/users/[id]/payment-plans/route.ts` - Query payment_plan_summary view

- `src/app/api/admin/payment-plans/process-manual/route.ts` - Testing endpoint

 

### Database

- `supabase/migrations/2025-11-06-refactor-payment-plans-to-xero-payments.sql` - Complete migration

 

### Tests

- `src/__tests__/services/payment-plan-service.test.ts` - **NEW** Rounding tests

- `src/__tests__/api/admin/user-payment-plans.test.ts` - Fixed mocking

- `src/__tests__/api/user/payment-plan-early-payoff.test.ts` - Updated for staging_metadata

 

## Breaking Changes

 

None - This is a backend refactor with no API contract changes. All existing endpoints continue to work with the same request/response formats.

 

## Deployment Notes

 

1. Run migration: `supabase migration up`

2. No data migration needed (development environment)

3. Production deployment would require careful data migration from old tables

 

## Related Issues

 

Addresses multiple Copilot review comments regarding:

- Code duplication

- Hardcoded values

- Null safety

- Performance optimization

- Schema mismatches

- Accounting accuracy

 

---

 

**All tests passing ✅** | **Ready for review**

### CODE REVIEW

File: src/app/api/stripe-webhook/route.ts [hunk 1]
```diff
@@ -1455,25 +1460,41 @@ export async function POST(request: NextRequest) {
...
+              .update({
+                payment_status: 'paid',
+                registered_at: new Date().toISOString(),
+              })
               .eq('user_id', paymentIntent.metadata.userId)
               .eq('registration_id', paymentIntent.metadata.registrationId)
-              .eq('payment_id', updatedPayment.id)
+              .in('payment_status', ['awaiting_payment', 'processing'])
+              .select()
               .single()
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/app/api/stripe-webhook/route.ts]: This is a good catch. This will be fixed in the next commit.

File: src/lib/xero/batch-sync-xero.ts [hunk 2]
```diff
@@ -613,6 +613,37 @@ export class XeroBatchSyncManager {
...
+          .from('payment_plans')
+          .select('next_payment_date, installments_paid, installments_count')
+          .eq('id', stagingMetadata.payment_plan_id)
+          .single()
+
+        if (paymentPlan) {
+          // Calculate final payment date: next_payment_date + (remaining installments * 30 days)
+          const remainingInstallments = paymentPlan.installments_count - paymentPlan.installments_paid
+          const nextPaymentDate = new Date(paymentPlan.next_payment_date)
+          const finalPaymentDate = new Date(nextPaymentDate.getTime() + (remainingInstallments - 1) * 30 * 24 * 60 * 60 * 1000)
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/xero/batch-sync-xero.ts]: claude committed code to address this
will re-run co-pilot review

File: src/lib/xero/batch-sync-xero.ts [hunk 3]
```diff
@@ -613,6 +613,37 @@ export class XeroBatchSyncManager {
...
+
+      if (stagingMetadata?.is_payment_plan && stagingMetadata?.payment_plan_id) {
+        // Fetch payment plan to get the final payment date
+        const { data: paymentPlan } = await this.supabase
+          .from('payment_plans')
+          .select('next_payment_date, installments_paid, installments_count')
+          .eq('id', stagingMetadata.payment_plan_id)
+          .single()
+
+        if (paymentPlan) {
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/xero/batch-sync-xero.ts]: claude committed code to address this
will re-run co-pilot review

File: src/lib/xero/batch-sync-xero.ts [hunk 4]
```diff
@@ -613,6 +617,64 @@ export class XeroBatchSyncManager {
...
+          //   - installments_paid = 1 (installment #1 completed)
+          //   - next_payment_date = Day 30 (installment #2)
+          //   - remaining after next = 2 (installments #3 and #4)
+          //   - final date = Day 30 + (2 * 30 days) = Day 90 (installment #4)
+          const remainingInstallments = paymentPlan.installments_count - paymentPlan.installments_paid
+          const nextPaymentDate = new Date(paymentPlan.next_payment_date)
+
+          // Remaining installments AFTER next_payment_date
+          const installmentsAfterNext = remainingInstallments - 1
+          const finalPaymentDate = new Date(nextPaymentDate.getTime() + (installmentsAfterNext * DAYS_30_IN_MS))
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/xero/batch-sync-xero.ts]: This is a valid point. However instead of doing this, i think it makes more sense to remove the payment_plans and related tables and instead put all of the information into xero_invoices and xero_payments. These tables can act as the single source of truth. This will ensure that all payments are properly staged in xero_payments as 'planned' status, and we can determine the last expected payment date with extra fields in xero_payments indicating the payment date.

File: src/lib/xero/batch-sync-xero.ts [hunk 5]
```diff
@@ -13,6 +13,10 @@ import * as Sentry from '@sentry/nextjs'
 import { getActiveTenant, validateXeroConnection } from './client'
 import { centsToCents, centsToDollars } from '../../types/currency'
 
+// Constants for date calculations
+const DAYS_30_IN_MS = 30 * 24 * 60 * 60 * 1000 // 30 days in milliseconds
+const DEFAULT_INVOICE_DUE_DAYS = 30 // Default due date for regular invoices
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/xero/batch-sync-xero.ts]: we will refactor the code to get the invoice due date based on the last payment record staged as 'planned' in xero_payments in a refactor to remove the payment_plans and payment_plans_transactions tables.

File: src/lib/services/payment-plan-service.ts [hunk 6]
```diff
@@ -744,18 +689,39 @@ export class PaymentPlanService {
...
-        .eq('user_id', userId)
-        .eq('status', 'active')
+      // Check if user has any planned or staged installment payments
+      const { data: payments, error } = await supabase
+        .from('xero_payments')
+        .select('id')
+        .eq('payment_type', 'installment')
+        .in('sync_status', ['planned', 'staged'])
+        .limit(1)
+
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/services/payment-plan-service.ts]: fixed by claude

File: src/lib/services/payment-plan-service.ts [hunk 7]
```diff
@@ -550,47 +505,39 @@ export class PaymentPlanService {
...
+          staging_metadata: {
+            ...(firstPlannedPayment?.staging_metadata || {}),
+            payment_id: paymentRecord.id,
+            early_payoff: true,
+            processed_at: new Date().toISOString()
+          }
         })
-        .eq('id', paymentPlanId)
+        .eq('xero_invoice_id', xeroInvoiceId)
+        .eq('sync_status', 'planned')
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/services/payment-plan-service.ts]: Good catch. What we'll do is mark all existing planned xero_payment records as cancelled, then insert a new staged record for the balance due.

File: src/lib/services/payment-plan-service.ts [hunk 8]
```diff
@@ -334,60 +299,39 @@ export class PaymentPlanService {
...
-        await adminSupabase
-          .from('payment_plans')
-          .update({
-            paid_amount: newPaidAmount,
-            installments_paid: newInstallmentsPaid,
-            status: isComplete ? 'completed' : 'active',
-            next_payment_date: nextPaymentDate
+            sync_status: 'pending',
+            staging_metadata: {
+              ...xeroPayment.staging_metadata,
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/services/payment-plan-service.ts]: will implement

File: src/lib/services/payment-plan-service.ts [hunk 9]
```diff
@@ -744,18 +698,31 @@ export class PaymentPlanService {
...
+        return false
+      }
+
+      const invoiceIds = invoices.map(inv => inv.id)
+
+      const { data: outstandingPayments, error: paymentsError } = await supabase
+        .from('xero_payments')
+        .select('id')
+        .in('xero_invoice_id', invoiceIds)
+        .in('sync_status', ['planned', 'staged'])
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/services/payment-plan-service.ts]: this logic is correct. outstanding balance is that which is not paid (planned or staged) but pending means it's been paid.

File: src/lib/services/payment-plan-service.ts [hunk 10]
```diff
@@ -266,66 +226,71 @@ export class PaymentPlanService {
...
       await adminSupabase
-        .from('payment_plan_transactions')
-        .update({ stripe_payment_intent_id: paymentIntent.id })
-        .eq('id', transactionId)
+        .from('xero_payments')
+        .update({
+          staging_metadata: {
+            ...xeroPayment.staging_metadata,
+            stripe_payment_intent_id: paymentIntent.id
+          }
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/services/payment-plan-service.ts]: fixed

File: src/app/api/stripe-webhook/route.ts [hunk 11]
```diff
@@ -147,6 +147,45 @@ const stripe = new Stripe(process.env.STRIPE_SECRET_KEY!, {
...
+
+    console.log('✅ Updated xero_payments statuses: #1=pending, #2-4=planned')
+  } else {
+    console.error(`❌ Expected exactly 4 installment payments, but found ${allPayments?.length || 0}`, {
+      xeroInvoiceId,
+      paymentCount: allPayments?.length || 0,
+      payments: allPayments
+    })
+  }
+}
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/app/api/stripe-webhook/route.ts]: good catch. we do not need to check for exactly 4 payments.

File: src/lib/services/payment-plan-service.ts [hunk 12]
```diff
@@ -79,125 +83,83 @@ export class PaymentPlanService {
...
-          amount: installmentAmount,
+        // For the last installment, absorb any rounding remainder to ensure exact total
+        const paymentAmount = i === PAYMENT_PLAN_INSTALLMENTS
+          ? data.totalAmount - (installmentAmount * (PAYMENT_PLAN_INSTALLMENTS - 1))
+          : installmentAmount
+
+        xeroPayments.push({
+          xero_invoice_id: data.xeroInvoiceId,
+          tenant_id: data.tenantId,
+          xero_payment_id: crypto.randomUUID(), // Placeholder, will be replaced when synced to Xero
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/services/payment-plan-service.ts]: fixing with 00000000-0000-0000-0000-000000000000 because column does not allow null

File: src/lib/services/payment-plan-service.ts [hunk 13]
```diff
@@ -497,20 +446,109 @@ export class PaymentPlanService {
...
+        .from('payments')
+        .insert({
+          user_id: userId,
+          total_amount: centsToCents(remainingBalance),
+          final_amount: centsToCents(remainingBalance),
+          status: 'pending', // Webhook will update to completed
+          payment_method: 'stripe'
+        })
+        .select()
+        .single()
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/services/payment-plan-service.ts]: will handle with more extensive logging

File: src/lib/services/payment-plan-service.ts [hunk 14]
```diff
@@ -79,125 +83,83 @@ export class PaymentPlanService {
...
-          },
-          'warn'
-        )
-        // Don't fail the whole operation - the plan is created
-      }
+      // Mark invoice as payment plan
+      await adminSupabase
+        .from('xero_invoices')
+        .update({ is_payment_plan: true })
+        .eq('id', data.xeroInvoiceId)
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/services/payment-plan-service.ts]: will fix

File: src/lib/xero/batch-sync-xero.ts [hunk 15]
```diff
@@ -613,6 +616,51 @@ export class XeroBatchSyncManager {
...
+          .eq('payment_type', 'installment')
+          .order('installment_number', { ascending: false })
+          .limit(1)
+          .single()
+
+        if (paymentError) {
+          // Database error querying xero_payments
+          console.error(`❌ Error fetching final payment date for payment plan invoice ${invoiceRecord.id}:`, paymentError)
+          console.warn('⚠️ Falling back to default 30-day due date due to database error')
+          dueDate = new Date(new Date(invoiceRecord.created_at).getTime() + DAYS_30_IN_MS).toISOString().split('T')[0]
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/lib/xero/batch-sync-xero.ts]: will implement

File: src/app/api/stripe-webhook/route.ts [hunk 16]
```diff
@@ -147,6 +147,53 @@ const stripe = new Stripe(process.env.STRIPE_SECRET_KEY!, {
...
+      .from('xero_payments')
+      .update({ sync_status: 'planned' })
+      .in('id', plannedPaymentIds)
+      .eq('sync_status', 'staged')
+  }
+
+  console.log(`✅ Updated xero_payments statuses: #1=pending, #${allPayments.length > 1 ? `2-${allPayments.length}` : '1 only'}=planned`, {
+    xeroInvoiceId,
+    installmentCount: allPayments.length
+  })
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply src/app/api/stripe-webhook/route.ts]: will fix

File: supabase/migrations/2025-11-06-refactor-payment-plans-to-xero-payments.sql [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+-- Grant access only to service_role (used by admin APIs and cron jobs)
+-- View uses security_invoker=true, so it still respects RLS on underlying tables
+GRANT SELECT ON payment_plan_summary TO service_role;
+
+-- 7. Drop old payment plan tables
+-- Skipping data migration since this is development environment
+DROP TABLE IF EXISTS payment_plan_transactions CASCADE;
+DROP TABLE IF EXISTS payment_plans CASCADE;
+
+-- 7. Add helpful comments
```
[the remark this answers is not in the document]
Comment (dwenderf) [reply supabase/migrations/2025-11-06-refactor-payment-plans-to-xero-payments.sql]: will fix
